beacon-chain/p2p/utils.go

```
@@ -123,43 +124,130 @@ func privKeyFromFile(path string) (*ecdsa.PrivateKey, error) {
+		Syncnets:  bitfield.NewBitvector4(),
+	}
+	wrappedDefaultMd := wrapper.WrappedMetadataV1(defaultMd)
 
-	_, err := os.Stat(defaultKeyPath)
-	defaultMetadataExist := !os.IsNotExist(err)
-	if err != nil && defaultMetadataExist {
-		return nil, err
+	// If --p2p-static-id is false, return default metadata for initialization
+	if !cfg.StaticPeerID {
+		return wrappedDefaultMd, nil
 	}
-	if metaDataPath == "" && !defaultMetadataExist {
-		metaData := &pb.MetaDataV0{
-			SeqNumber: 0,
-			Attnets:   bitfield.NewBitvector64(),
-		}
-		dst, err := proto.Marshal(metaData)
+
+	mdPath, exist := resolveMetaDataPath(cfg)
+	if exist {
+		md, err := metaDataFromFile(mdPath)
 		if err != nil {
+			if errors.Is(err, ssz.ErrSize) {
+				// In case previous metadata file is encoded by proto,
+				// we need to migrate it into ssz encoded version.
+				return migrateFromProtoToSsz(mdPath)
+			}
 			return nil, err
 		}
-		if err := file.WriteFile(defaultKeyPath, dst); err != nil {
-			return nil, err
+		return md, err
+	}
+	if err := saveMetaDataToFile(mdPath, wrappedDefaultMd); err != nil {
+		return nil, err
+	}
+
+	return wrappedDefaultMd, nil
+}
+
+// resolveMetaDataPath returns path and the existence of that path.
+// Issue while opening a file(e.g. permission issues) will be handled at metaDataFromFile.
+func resolveMetaDataPath(cfg *Config) (string, bool) {
+	var mdPath string
+
+	// Prioritize if --p2p-metadata is provided.
+	if cfg.MetaDataDir != "" {
+		mdPath = cfg.MetaDataDir
+	} else {
+		mdPath = path.Join(cfg.DataDir, metaDataPath)
+	}
+
+	// Return path and existence of the file.
+	_, err := os.Stat(mdPath)
+	if !os.IsNotExist(err) {
+		return mdPath, true
+	}
+	return mdPath, false
+}
+
+// metaDataFromFile retrieves unmarshalled p2p metadata from file.
+func metaDataFromFile(path string) (metadata.Metadata, error) {
+	src, err := os.ReadFile(path) // #nosec G304
+	if err != nil {
+		log.WithError(err).Error("Error reading metadata from file")
```

No need to log here. You can return a wrapped error instead.